File: benchmarks/pandas/bench_cat_union_intersect_diff.py
import pandas as pd, time, json
N = 50_000
cats1 = ["A", "B", "C", "D"]
cats2 = ["C", "D", "E", "F"]
s1 = pd.Categorical([cats1[i % len(cats1)] for i in range(N)], categories=cats1)
s2 = pd.Categorical([cats2[i % len(cats2)] for i in range(N)], categories=cats2)
WARMUP = 3
ITERS = 20
for _ in range(WARMUP):
    _ = s1.set_categories(s1.categories.union(s2.categories))
    _ = s1.set_categories(s1.categories.intersection(s2.categories))
    _ = s1.set_categories(s1.categories.difference(s2.categories))
t0 = time.perf_counter()
for _ in range(ITERS):
    _ = s1.set_categories(s1.categories.union(s2.categories))
    _ = s1.set_categories(s1.categories.intersection(s2.categories))
    _ = s1.set_categories(s1.categories.difference(s2.categories))
total = (time.perf_counter() - t0) * 1000
print(json.dumps({"function": "cat_union_intersect_diff", "mean_ms": total / ITERS, "iterations": ITERS, "total_ms": total}))

File: benchmarks/tsb/bench_cat_union_intersect_diff.ts
import { Series, catUnionCategories, catIntersectCategories, catDiffCategories } from "tsb";
const N = 50_000;
const cats1 = ["A", "B", "C", "D"];
const cats2 = ["C", "D", "E", "F"];
const s1 = new Series(Array.from({ length: N }, (_, i) => cats1[i % cats1.length]));
const s2 = new Series(Array.from({ length: N }, (_, i) => cats2[i % cats2.length]));
const c1 = s1.cat;
const c2 = s2.cat;
const WARMUP = 3;
const ITERS = 20;
for (let i = 0; i < WARMUP; i++) {
  catUnionCategories(c1, c2);
  catIntersectCategories(c1, c2);
  catDiffCategories(c1, c2);
}
const t0 = performance.now();
for (let i = 0; i < ITERS; i++) {
  catUnionCategories(c1, c2);
  catIntersectCategories(c1, c2);
  catDiffCategories(c1, c2);
}
const total = performance.now() - t0;
console.log(JSON.stringify({ function: "cat_union_intersect_diff", mean_ms: total / ITERS, iterations: ITERS, total_ms: total }));
